### ISSUE TITLE
ipvs proxier wrongly unbind addr when multiple services have the same external IP

### ISSUE THREAD

**Issue by ogre0403** [kubernetes/kubernetes#80158]
**What happened**:

In kube-proxy IPVS mode, kube-proxy will unbind legacy address from kue-ipvs0. If service has external IP, external IP is also unbind from kube-ipvs0.  In our use case, services in the same namespace share ONE external IP. 
```
tx9hp5    agwg8wvggsc2gpu           LoadBalancer   10.105.113.46    172.29.188.51    59417:30594/TCP,56220:30784/TCP   4d
tx9hp5    alp4p0shortshirt1gpu      LoadBalancer   10.111.157.13    172.29.188.51    58527:31385/TCP,55107:32030/TCP   3h
tx9hp5    axusz5hash32bit1gpu       LoadBalancer   10.104.63.214    172.29.188.51    57424:32482/TCP,52964:31991/TCP   3h
```
When we delete one svc, the external IP is unbind from kube-ipvs0, then other services use the same external IP will encounter broken connection.  
However, for the next round of sync, the external IＰis bind to kube-ipvs0 again. 

**What you expected to happen**:

If some service still have same IPs, the IP should not be unbind. 

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 
1.10.10
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
CentOS 7
- Kernel (e.g. `uname -a`):
3.10.0-862.el7.x86_64 

- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by thockin** [comment]
What environment is setting the same IP for multiple services?  Is this a custom "provider" for LBs?

It sounds like a bug in IPVS.  @andrewsykim @m1093782566

**Comment by andrewsykim** [comment]
yeah it seems like the IPVS proxier assumes that if an external IP is removed from a Service it will delete any associated connections to it without checking if the same externalIP is used in another service. See [deleteEndpointConnections](https://github.com/kubernetes/kubernetes/blob/c25763e159a4560b6a7b794c109255eebbf6385d/pkg/proxy/ipvs/proxier.go#L1554-L1579).

**Comment by andrewsykim** [comment]
But this should only apply for UDP services as far as I can tell, @ogre0403 can you confirm if those external IPs are using UDP? 

/assign

**Comment by thockin** [comment]
Why only UDP?  In theory, any LB IP could be shared if ports are disjoint.

**Comment by andrewsykim** [comment]
> Why only UDP?  In theory, any LB IP could be shared if ports are disjoint.

Ah good point. Also realizing now that the logic for cleaning up stale UDP connections does account for servie ports so my previous theory is incorrect. 

The IPVS proxier does account for external IPs on multiple services though - if it sees at least 1 service using that external IP it should acknowledge that and not unbind it. The only scenario I see this happening is if the externalIP check fails for every other service using the externalIP so the proxier thinks it should delete it. @ogre0403 are you able to share the service spec for all the services using that externalIP?

**Comment by ogre0403** [comment]
We think this issue has been resolved in v1.12.  

We've updated our kube-proxy to v1.12.5. In this version, proxier  will find out the legacy bind address first, then delete only  these addresses.  

https://github.com/kubernetes/kubernetes/blob/v1.12.5/pkg/proxy/ipvs/proxier.go#L1658

```
func (proxier *Proxier) getLegacyBindAddr(activeBindAddrs map[string]bool, currentBindAddrs []string) map[string]bool {
	legacyAddrs := make(map[string]bool)
	for _, addr := range currentBindAddrs {
		if _, ok := activeBindAddrs[addr]; !ok {
			legacyAddrs[addr] = true
		}
	}
	return legacyAddrs
}
```

**Comment by andrewsykim** [comment]
Ha, didn't realize this was for kube-proxy v1.10.10 😬 

Feel free to re-open if you run into the same issue in v1.12 

/close
